File: examples/03-retry-catch/main.go
// Example 03: retry policy + catch state — flaky external API call
// FetchExternalData (retry ×3) → EnrichRecord → Persist → end
//
//	↓ catch
//
// HandleFetchError → end
package main

import (
	"fmt"
	"log"

	"github.com/pastorenue/kflow/examples/03-retry-catch/handlers"
	"github.com/pastorenue/kflow/internal/local"
	"github.com/pastorenue/kflow/pkg/kflow"
)

func buildWorkflow(h *handlers.PipelineHandlers) *kflow.Workflow {
	wf := kflow.New("data-pipeline")

	wf.Task("FetchExternalData", h.FetchExternalData).
		Retry(kflow.RetryPolicy{MaxAttempts: 3, BackoffSeconds: 0}).
		Catch("HandleFetchError")

	wf.Task("EnrichRecord", h.EnrichRecord)
	wf.Task("Persist", h.Persist)
	wf.Task("HandleFetchError", h.HandleFetchError)

	wf.Flow(
		kflow.Step("FetchExternalData").Next("EnrichRecord").Catch("HandleFetchError"),
		kflow.Step("EnrichRecord").Next("Persist"),
		kflow.Step("Persist").End(),
		kflow.Step("HandleFetchError").End(),
	)
	return wf
}

func main() {
	fmt.Println("=== 03-retry-catch: data-pipeline ===")
	if err := local.RunLocal(buildWorkflow(handlers.New(2)), kflow.Input{"record_id": "REC-9900"}); err != nil {
		log.Fatalf("workflow failed: %v", err)
	}
	fmt.Println("  → workflow COMPLETED")

	fmt.Println("\n--- triggering catch (all retries exhausted) ---")
	if err := local.RunLocal(buildWorkflow(handlers.New(0)), kflow.Input{"record_id": "REC-9901"}); err != nil {
		log.Fatalf("workflow failed: %v", err)
	}
	fmt.Println("  → workflow COMPLETED (via catch)")
}

File: examples/03-retry-catch/python/main.py
# Example 03: retry policy + catch state — flaky external API call
# FetchExternalData (retry ×3) → EnrichRecord → Persist → end
#                ↓ catch
# HandleFetchError → end
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../../../sdk/python"))
import kflow

from handlers import (
    STATE_FETCH_EXTERNAL_DATA,
    STATE_ENRICH_RECORD,
    STATE_PERSIST,
    STATE_HANDLE_FETCH_ERROR,
    PipelineHandlers,
)


def build_workflow(h: PipelineHandlers):
    wf = kflow.Workflow("data-pipeline")
    wf.task(STATE_FETCH_EXTERNAL_DATA)(h.fetch_external_data)
    wf.task(STATE_ENRICH_RECORD)(h.enrich_record)
    wf.task(STATE_PERSIST)(h.persist)
    wf.task(STATE_HANDLE_FETCH_ERROR)(h.handle_fetch_error)

    wf.flow(
        kflow.step(STATE_FETCH_EXTERNAL_DATA)
            .next(STATE_ENRICH_RECORD)
            .catch(STATE_HANDLE_FETCH_ERROR)
            .retry(3, 0),
        kflow.step(STATE_ENRICH_RECORD).next(STATE_PERSIST),
        kflow.step(STATE_PERSIST).end(),
        kflow.step(STATE_HANDLE_FETCH_ERROR).end(),
    )
    return wf


def main():
    print("=== 03-retry-catch: data-pipeline ===")
    kflow.run_local(build_workflow(PipelineHandlers(succeed_after=2)), {"record_id": "REC-9900"})
    print("  → workflow COMPLETED")

    print("\n--- triggering catch (all retries exhausted) ---")
    kflow.run_local(build_workflow(PipelineHandlers(succeed_after=0)), {"record_id": "REC-9901"})
    print("  → workflow COMPLETED (via catch)")


if __name__ == "__main__":
    main()
